src/Microsoft.IdentityModel.Tokens/JsonWebKeySet.cs

```diff
@@ -122,68 +120,100 @@ public JsonWebKeySet(string json, JsonSerializerSettings jsonSerializerSettings)
 
-                if (!StringComparer.Ordinal.Equals(webKey.Kty, JsonWebAlgorithmsKeyTypes.RSA))
+            foreach (var webKey in Keys)
+            {
+                // skip if "use" (Public Key Use) parameter is not empty or "sig"
+                if (!(string.IsNullOrWhiteSpace(webKey.Use) || webKey.Use.Equals(JsonWebKeyUseNames.Sig, StringComparison.Ordinal)))
+                {
+                    LogHelper.LogInformation(LogHelper.FormatInvariant(LogMessages.IDX10808, webKey.Use));
                     continue;
+                }
 
-                if ((string.IsNullOrWhiteSpace(webKey.Use) || (StringComparer.Ordinal.Equals(webKey.Use, JsonWebKeyUseNames.Sig))))
+                if (webKey.Kty.Equals(JsonWebAlgorithmsKeyTypes.RSA, StringComparison.Ordinal))
                 {
                     if (webKey.X5c != null)
-                    {
-                        foreach (var certString in webKey.X5c)
-                        {
-                            try
-                            {
-                                // Add chaining
-                                SecurityKey key = new X509SecurityKey(new X509Certificate2(Convert.FromBase64String(certString)));
-                                key.KeyId = webKey.Kid;
-                                keys.Add(key);
-                            }
-                            catch (CryptographicException ex)
-                            {
-                                throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10802, webKey.X5c[0]), ex));
-                            }
-                            catch (FormatException fex)
-                            {
-                                throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10802, webKey.X5c[0]), fex));
-                            }
-                        }
-                    }
+                        signingKeys.AddRange(CreateX509SecurityKeys(webKey));
 
                     if (!string.IsNullOrWhiteSpace(webKey.E) && !string.IsNullOrWhiteSpace(webKey.N))
+                        signingKeys.Add(CreateRsaSecurityKey(webKey));
+                }
+                else if (webKey.Kty.Equals(JsonWebAlgorithmsKeyTypes.EllipticCurve, StringComparison.Ordinal))
+                {
+                    signingKeys.Add(CreateECDsaSecurityKey(webKey));
+                }
+                else
+                {
+                    //kty is not 'EC' or 'RSA'
+                    LogHelper.LogInformation(LogHelper.FormatInvariant(LogMessages.IDX10809, webKey.Kty));
+                }
+            }
+
+            return signingKeys;
+        }
+
+        private IList<SecurityKey> CreateX509SecurityKeys(JsonWebKey jsonWebKey)
+        {
+            try
+            {
+                var keys = new List<SecurityKey>();
+
+                foreach (var certString in jsonWebKey.X5c)
```

Comment:
It seems from the spec, that only the first key is used to perform any signing operations. The other keys are used to supply chain verification.

I am not sure we want to add them all.

Have a look at this: https://tools.ietf.org/html/rfc7517#section-4.7

Reply:
I agree. Opened a new issue to track this: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1146